internal/controller/customreplicaset_controller.go

```diff
@@ -74,13 +80,130 @@ func (r *CustomReplicaSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	availableStdPods, availableUpgPods := countAvailablePods(childPods.Items)
 	totalAvailablePods := availableStdPods + availableUpgPods
 
+	// Load latest controller revision
+	var latestRevision *v1.ControllerRevision
+	if latestRevision, err = r.getLatestControllerRevision(ctx, &crs); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Apply any necessary updates to controller revision history
+	if err := r.manageControllerRevisionHistory(ctx, &crs, latestRevision); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
 	if err := r.managePods(ctx, totalAvailablePods, availableStdPods, availableUpgPods, &crs, childPods, log); err != nil {
 		return ctrl.Result{Requeue: true}, err
 	}
 
 	return ctrl.Result{Requeue: false}, err
 }
 
+// Create a new ControllerRevision
+func (r *CustomReplicaSetReconciler) createControllerRevision(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, latestRevision *v1.ControllerRevision, jsonSpec []byte) error {
+	var revisionNumber int64
+	if latestRevision == nil {
+		revisionNumber = 0
+	} else {
+		revisionNumber = latestRevision.Revision + 1
+	}
+
+	t := time.Now()
+	timestamp := fmt.Sprintf("%d%d%d%d", t.Hour(), t.Minute(), t.Second(), t.Nanosecond())
+
+	newRevision := v1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-controller-revision-" + timestamp,
+			Namespace: cr.Namespace,
+			Labels:    cr.Labels,
+		},
+		Data:     runtime.RawExtension{Raw: jsonSpec},
+		Revision: revisionNumber,
+	}
+	if err := controllerutil.SetControllerReference(cr, &newRevision, r.Scheme); err != nil {
+		return err
+	}
+	if err := r.Create(ctx, &newRevision); err != nil {
+		return err
+	}
+
+	// Update 'revisionNumber' label of the CustomReplicaSet
+	if cr.Labels == nil {
+		cr.Labels = map[string]string{}
+	}
+	cr.Labels["revisionNumber"] = strconv.FormatInt(newRevision.Revision, 10)
+	if err := r.Update(ctx, cr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *CustomReplicaSetReconciler) manageControllerRevisionHistory(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, latestRevision *v1.ControllerRevision) error {
+	if latestRevision == nil {
+		jsonSpec, err := convertSpecToJson(cr)
+		if err != nil {
+			return err
+		}
+
+		if err := r.createControllerRevision(ctx, cr, latestRevision, jsonSpec); err != nil {
+			fmt.Println("Failed to create first controller revision")
+			return err
+		}
+		return nil
+	}
+	var latestSpec customreplicasetv1.CustomReplicaSetSpec
+
+	if err := json.Unmarshal(latestRevision.Data.Raw, &latestSpec); err != nil {
+		fmt.Println("Failed to unmarshal latest controller spec to json")
+		return err
+	}
+
+	// Compare the PodTemplates
+	if !reflect.DeepEqual(cr.Spec.Template, latestSpec.Template) {
+		jsonSpec, err := convertSpecToJson(cr)
```

Comment:
lines 163-171 and lines 143-152 are mostly identical. Try to refactor code to avoid duplication.

Reply:
Yes cleaned this up to use hash method instead